### DISCUSSION TITLE
How can I decrease workflow definition storage?

### DISCUSSION THREAD

**Discussion by amirpopovich** [Netflix/conductor#2882] [Q&A] [answered]
Hi, 

I'm looking at conductor to be a scalable workflow solution, aiming for over 1m flows per day.

I can see that the runtime persists the workflow definition object on each workflow instance that I start.
My definitions are pretty big, and can easily get to 100kb.

The workflow definition & version both appear on the MetadataDAO persistence level.

Questions:
1. What was the reason of adding the definition to each workflow instance?
2. Is there anyway to avoid saving the definition on each workflow instance and only save the definitionId and it's version on each instance while reading the definition from the metadataDAO when needed?

Thank you.

**Comment by apanicker-nflx** [comment, marked as answer]
> What was the reason of adding the definition to each workflow instance?

This was done to lock in a workflow execution with the workflow definition that the workflow was launched with.

> Is there anyway to avoid saving the definition on each workflow instance and only save the definitionId and it's version on each instance while reading the definition from the metadataDAO when needed?

This is not possible currently, because automatic versioning is not enforced on the workflow definition. So a workflow can be launched with a workflow definition at version x. A user can then update this workflow definition and keep the version as x. The workflow execution now in progress can be deemed invalid as a result of this change and the execution would be left in an unstable hung state.

**Reply by apanicker-nflx** [reply]
> My definitions are pretty big, and can easily get to 100kb.

Is each individual workflow definition expected to be ~100kb. Curious to know what types of workflows you are running?

**Comment by amirpopovich** [comment]
@apanicker-nflx 

Thank you for your reply.
Regarding the 100kb flows, we are talking about large workflows with around 60-70 tasks that build up to pretty large flows.

The issue with the definition length is mainly important to us since we thought about using Redis as our main DAO, and Redis doesn't like large objects due to it's single threaded runtime.
Is enforcing automatic versioning somewhere on your roadmap?

**Reply by apanicker-nflx** [reply]
We introduced metadata mutability in the UI in [v3.6.0](https://github.com/Netflix/conductor/releases/tag/v3.6.0). As part of this change, the UI defaults to auto-incrementing versions but this can be overridden. Eventually, we would want to enforce automatic versioning but the main challenge lies in doing this in a manner that is least intrusive to our current users. This is on our long term roadmap but won't be picked up in the near future.
